### PR TITLE
[Kotlin][Client] Added supporting Retrofit2(RxJava/RxJava2/Coroutines) 

### PR DESCRIPTION
**Pull request by kuFEAR**
Added supporting Retrofit2(RxJava/RxJava2/Coroutines) for Kotlin client code generator

### PR checklist

- [x] Read the [contribution guidelines](https://github.com/openapitools/openapi-generator/blob/master/CONTRIBUTING.md).
- [x] If contributing template-only or documentation-only changes which will change sample output, [build the project](https://github.com/OpenAPITools/openapi-generator#14---build-projects) before.
- [x] Run the shell script(s) under `./bin/` (or Windows batch scripts under`.\bin\windows`) to update Petstore samples related to your fix. This is important, as CI jobs will verify _all_ generator outputs of your HEAD commit, and these must match the expectations made by your contribution. You only need to run `./bin/{LANG}-petstore.sh`, `./bin/openapi3/{LANG}-petstore.sh` if updating the code or mustache templates for a language (`{LANG}`) (e.g. php, ruby, python, etc).
- [x] File the PR against the [correct branch](https://github.com/OpenAPITools/openapi-generator/wiki/Git-Branches): `master`, `4.3.x`, `5.0.x`. Default: `master`.
- [x] Copy the [technical committee](https://github.com/openapitools/openapi-generator/#62---openapi-generator-technical-committee) to review the pull request if your PR is targeting a particular programming language.

@jimschubert @dr4ke616 @karismann @Zomzog @andrewemery @4brunu

### CODE REVIEW

**Comment by wing328** [comment]
Please merge the latest master into this branch to resolve the CircleCI failure.

**Comment by kuFEAR** [comment]
hm, merge is not helped

**Comment by 4brunu** [comment]
Currently we have two open PR that implement coroutines support.
https://github.com/OpenAPITools/openapi-generator/pull/5697
@wing328 could you please help with this? How should we handle this?

**Comment by wing328** [comment]
I saw a couple of build failures in https://github.com/OpenAPITools/openapi-generator/pull/5697 and looks like this one is ready to merge.

@Ch4rl3x I wonder if you can also review this PR to see if it meets your requirement.

**Comment by Ch4rl3x** [comment]
> I saw a couple of build failures in #5697 and looks like this one is ready to merge.
> 
> @Ch4rl3x I wonder if you can also review this PR to see if it meets your requirement.

I checked the PR. Supend call is added but the ApiClient in Infrastructure is still minimalistic and not comparable with the java-retrofit2 ApiClient. This PR has some overlaps with our PR. Maybe i could pull from master again after this PR is merged and change my PR to add the missing stuff

**Comment by Ch4rl3x** [comment]
Maybe this PR should be merged and later on i will add the missing features from #5697 in a new PR? Any other ideas how to solve these conflict and get progress?

**Comment by wing328** [comment]
Let me run some tests locally and merge if no issue found.

**Comment by wing328** [comment]
I did some tests locally with the kotlin-related samples under `samples/openapi3/client/petstore` using `gradle build` but got a lot of errors

Does it work for you locally?

**Comment by Ch4rl3x** [comment]
Could you make a few tests locally for #5697 also?
Maybe there are less errors. Both PRs should be merged in the future, maybe we should merge #5697 first and then add the features from #5750 ?

**Comment by wing328** [comment]
Saw your PR too but you renamed all the output directory, which makes it hard to review the actual change you want to make (~760 files changed...)

Understood your intention, which is good but still prefers to stick with the current naming convention of the output directory for the time being.

**Comment by kuFEAR** [comment]
> I did some tests locally with the kotlin-related samples under `samples/openapi3/client/petstore` using `gradle build` but got a lot of errors
> 
> Does it work for you locally?

As I noticed the problem https://github.com/OpenAPITools/openapi-generator/pull/5750#discussion_r402575012 I don't touched this part of generator(Kotlin/Gson), I think the generator Initially doesn't work properly, you can check this in samples/openapi3/client/petstore/kotlin/org.openapitools.client.models.NullableClass

what the problem I saw:
1. @SerializableName added for field in polymorphic model interface(org.openapitools.client.models.Animal), it should be add only in child classes of it(org.openapitools.client.models.Cat, org.openapitools.client.models.Dog) and this part works correct. [FIXED]
2. @SerializableName(value=1) should be written as @SerializableName(value="1") EnumTest.EnumInteger and EnumTest.EnumNumber [FIXED] also for jackson
3. https://github.com/OpenAPITools/openapi-generator/pull/5750#discussion_r402575012 kotlin.mapOf&lt;String, kotlin.Any&gt; it's strange mapping, there should be use just kotlin.collections.HashMap<String, Any>, it doesn't work on previous sample project org.openapitools.client.models.NullableClass [FIX 1/2] &lt;String, kotlin.Any&gt; escaping still is problem

**Comment by kuFEAR** [comment]
@wing328 do you know how to fix https://github.com/OpenAPITools/openapi-generator/pull/5750#issuecomment-626663248 (3) problem with escaping?

**Comment by kuFEAR** [comment]
For inherit of HashMap/ArrayList we need add empty braces, how I can determine that is abstract/non final class/interface in *.mustaches?

**Comment by kuFEAR** [comment]
https://github.com/OpenAPITools/openapi-generator/issues/5600 Fixed encoding from 3. using {{{parent}}}

**Comment by kuFEAR** [comment]
3. [FIXED] 
I fixed adding empty braces for inheritance from map/array by {{#isMapModel}}(){{/isMapModel}}{{#isArrayModel}}(){{/isArrayModel}}, I don't know is correct approach or not, but it works)

File: modules/openapi-generator/src/main/resources/kotlin-client/interface_opt_var.mustache [hunk 1]
```diff
@@ -5,9 +5,6 @@
     {{#moshi}}
     @Json(name = "{{{vendorExtensions.x-base-name-literal}}}")
     {{/moshi}}
-    {{#gson}}
```
Comment (4brunu) [inline modules/openapi-generator/src/main/resources/kotlin-client/interface_opt_var.mustache:8]: @kuFEAR Why did you remove gson from here?

File: modules/openapi-generator/src/main/resources/kotlin-client/interface_req_var.mustache [hunk 2]
```diff
@@ -5,9 +5,6 @@
     {{#moshi}}
     @Json(name = "{{{vendorExtensions.x-base-name-literal}}}")
     {{/moshi}}
-    {{#gson}}
```
Comment (4brunu) [inline modules/openapi-generator/src/main/resources/kotlin-client/interface_req_var.mustache:8]: @kuFEAR same here

Comment (kuFEAR) [reply modules/openapi-generator/src/main/resources/kotlin-client/interface_req_var.mustache:8]: https://github.com/OpenAPITools/openapi-generator/pull/5750#issuecomment-626663248 1.

File: modules/openapi-generator/src/main/resources/kotlin-client/interface_opt_var.mustache (hunk 1 above)
Comment (kuFEAR) [reply modules/openapi-generator/src/main/resources/kotlin-client/interface_opt_var.mustache:8]: https://github.com/OpenAPITools/openapi-generator/pull/5750#issuecomment-626663248 1.

Comment (kuFEAR) [reply modules/openapi-generator/src/main/resources/kotlin-client/interface_opt_var.mustache:8]: Gson @SerializedName target only for method and fields 
@Target({ElementType.FIELD, ElementType.METHOD}) and Kotlin interface fields is not yet determined as field by kotlin compiler, I saw error message there
Screenshot 2020-05-12 at 12 20 06

Comment (4brunu) [reply modules/openapi-generator/src/main/resources/kotlin-client/interface_opt_var.mustache:8]: Thanks

File: modules/openapi-generator/src/main/resources/kotlin-client/interface_req_var.mustache (hunk 2 above)
Comment (4brunu) [reply modules/openapi-generator/src/main/resources/kotlin-client/interface_req_var.mustache:8]: Thanks

Comment (kuFEAR) [reply modules/openapi-generator/src/main/resources/kotlin-client/interface_req_var.mustache:8]: I have an idea add @get:SerializedName("field")

Comment (4brunu) [reply modules/openapi-generator/src/main/resources/kotlin-client/interface_req_var.mustache:8]: If that works, it would be nice, because I use Gson on my projects

Comment (kuFEAR) [reply modules/openapi-generator/src/main/resources/kotlin-client/interface_req_var.mustache:8]: \+

Comment (4brunu) [reply modules/openapi-generator/src/main/resources/kotlin-client/interface_req_var.mustache:8]: Thanks

**Comment by kuFEAR** [comment]
```
[main] ERROR org.openapitools.codegen.languages.MysqlSchemaCodegen - Invalid database name. '12345' cannot be used as MySQL identifier. Escaped value '12345_db' will be used instead.
[main] ERROR org.openapitools.codegen.DefaultCodegen - `null` (map property) does not have a proper inner type defined. Default to type:string
[main] ERROR org.openapitools.codegen.DefaultCodegen - ERROR! Not handling  class QueryParameter {
    class Parameter {
        name: limit
        in: null
        description: this is a description
        required: true
        deprecated: null
        allowEmptyValue: null
        style: null
        explode: null
        allowReserved: null
        schema: null
        examples: null
        example: null
        content: null
        $ref: null
    }
    in: query
} as Body Parameter at the moment
[main] ERROR org.openapitools.codegen.DefaultCodegen - `null` (map property) does not have a proper inner type defined. Default to type:string
[main] ERROR org.openapitools.codegen.DefaultCodegen - body in fromRequestBody cannot be null!
[main] ERROR org.openapitools.codegen.utils.ModelUtils - Schema cannot be null in isFreeFormObject check
[main] ERROR o.o.c.config.CodegenConfigurator - Unexpected character ('-' (code 45)) in numeric value: expected digit (0-9) to follow minus sign, for valid numeric value
 at [Source: (File); line: 1, column: 3]
[main] ERROR o.o.c.config.CodegenConfigurator - Unexpected character ('-' (code 45)) in numeric value: expected digit (0-9) to follow minus sign, for valid numeric value
 at [Source: (File); line: 1, column: 3]
```

```
[ERROR] Failed to execute goal org.codehaus.mojo:exec-maven-plugin:1.2.1:exec (test) on project PythonFlaskConnexionTests: Command execution failed. Process exited with an error: 2 (Exit value: 2) -> [Help 1]
[ERROR] 
[ERROR] To see the full stack trace of the errors, re-run Maven with the -e switch.
[ERROR] Re-run Maven using the -X switch to enable full debug logging.
[ERROR] 
[ERROR] For more information about the errors and possible solutions, please read the following articles:
[ERROR] [Help 1] http://cwiki.apache.org/confluence/display/MAVEN/MojoExecutionException
[ERROR] 
[ERROR] After correcting the problems, you can resume the build with the command
[ERROR]   mvn <args> -rf :PythonFlaskConnexionTests
```

This error is related to my PR?

**Comment by 4brunu** [comment]
I don't think so, it seems to be related to python, and the master branch was failing with the same error.

**Comment by kuFEAR** [comment]
> I don't think so, it seems to be related to python, and the master branch was failing with the same error.

ok, thanks

**Comment by wing328** [comment]
The issue has been fixed in the master. Please merge the latest master into this branch to fix the issue.

**Comment by kuFEAR** [comment]
now it fails on sudo apt-get install elixir

**Comment by wing328** [comment]
Restarted the CI job. Let's see how it goes.

**Comment by wing328** [comment]
Tested the kotlin retrofit2 samples and no longer encounter the issue.
